src/puppetlabs/dujour/version_check.clj

```diff
@@ -0,0 +1,121 @@
+  [group-id artifact-id]
+  (version/get-version group-id artifact-id))
+
+(def version
+  "Get the version number of this installation."
+  (memoize version*))
+
+(schema/defn ^:always-validate update-info :- (schema/maybe UpdateInfo)
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Returns the JSON object received from the server, which
+  is expected to be a map containing keys `:version`, `:newer`, and `:link`.
+  Returns `nil` if the request does not succeed for some reason."
+  [product-name :- ProductName
+   update-server-url :- schema/Str]
+  (let [{:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (version group-id artifact-id)
+        version-data {:version current-version}
+        query-string (ring-codec/form-encode version-data)
+        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
+        {:keys [status body] :as resp} (client/get url
+                                                   {:headers {"Accept" "application/json"}
+                                                    :as :text})]
+    (if (= status 200)
+      (json/parse-string body true)
+      (sling/throw+ {:type ::update-request-failed
+                     :message resp}))))
+
+
+(defn validate-config!
+  [product-name update-server-url]
+  ;; if this ends up surfacing error messages that aren't very user-friendly,
+  ;; we can improve the validation logic.
+  (schema/validate ProductName product-name)
+  (schema/validate (schema/maybe schema/Str) update-server-url))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(defn check-for-updates!
+  "This will fetch the latest version number and log if the system
+  is out of date."
+  [product-name update-server-url]
+  (log/debugf "Checking for newer versions of %s" product-name)
+  (let [update-server-url             (or update-server-url default-update-server-url)
+        {:keys [version newer link]}  (try
+                                        (update-info product-name update-server-url)
+                                        (catch Throwable e
+                                          (log/debug e (format "Could not retrieve update information (%s)" update-server-url))))
+        link-str (if link
+                   (format " Visit %s for details." link)
+                   "")
+        update-msg (format "Newer version %s is available!%s" version link-str)]
+    (when newer
+      (log/info update-msg))
+    {}))
+
+(defn version-check
+  ([product-name update-server-url]
+    (version-check product-name update-server-url nil nil))
+  ([product-name update-server-url watch-key watch-fn]
+    (validate-config! product-name update-server-url)
+    (let [version-agent (agent {:status :in-progress})]
+      (if (and watch-key watch-fn)
+        (add-watch version-agent watch-key watch-fn))
+      (send version-agent
+            (try
```

Comment:
Ohhh, this works because we *are* given `send` a function - the empty `{}`!  Seems obvious now, but the `try` is evaluated once and just so happens to return something that is a function (the `{}`), which would explain you weren't seeing the state change and getting the Connection failures.